test/e2e/framework/metrics/metrics_grabber.go

```diff
@@ -235,16 +238,26 @@ func (g *Grabber) Grab() (Collection, error) {
 	return result, nil
 }
 
-func (g *Grabber) getMetricsFromPod(client clientset.Interface, podName string, namespace string, port int) (string, error) {
-	rawOutput, err := client.CoreV1().RESTClient().Get().
-		Namespace(namespace).
-		Resource("pods").
-		SubResource("proxy").
-		Name(fmt.Sprintf("%v:%v", podName, port)).
-		Suffix("metrics").
-		Do(context.TODO()).Raw()
-	if err != nil {
+func (g *Grabber) getMetricsFromPod(client clientset.Interface, podName string, namespace string, port int) (metrics string, err error) {
+	pollErr := wait.PollImmediate(5*time.Second, 2*time.Minute, func() (bool, error) {
```

Comment:
I made up the numbers based on the occurrence I've analyzed, feel free to suggest others

Reply:
I don't think it's a good idea to make this low-level helper method take up to 2 minutes to return. I'd rather see us stop using the GrabFromControllerManager() metrics collector in a way that will fail tests if the controller-manager pod isn't up at exactly that moment. That is not even guaranteed to be available in all environments.

An alternative would be to make the metrics grabber wait once for the k-c-m pod to be available in GrabFromControllerManager()

```diff
diff --git a/test/e2e/framework/metrics/metrics_grabber.go b/test/e2e/framework/metrics/metrics_grabber.go
index fdfd51f9997..ce41294cdbe 100644
--- a/test/e2e/framework/metrics/metrics_grabber.go
+++ b/test/e2e/framework/metrics/metrics_grabber.go
@@ -19,6 +19,7 @@ package metrics
 import (
        "context"
        "fmt"
+       "sync"
 
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
        "k8s.io/apimachinery/pkg/fields"
@@ -49,6 +50,8 @@ type Grabber struct {
        grabFromClusterAutoscaler bool
        masterName                string
        registeredMaster          bool
+
+       waitForControllerManagerReadyOnce sync.Once
 }
 
 // NewMetricsGrabber returns new metrics which are initialized.
@@ -160,6 +163,9 @@ func (g *Grabber) GrabFromControllerManager() (ControllerManagerMetrics, error)
        if !g.registeredMaster {
                return ControllerManagerMetrics{}, fmt.Errorf("Master's Kubelet is not registered. Skipping ControllerManager's metrics gathering")
        }
+       g.waitForControllerManagerReadyOnce.Do(func() {
+               // ... wait for pod to exist and be ready up to timeout ...
+       })
        output, err := g.getMetricsFromPod(g.client, fmt.Sprintf("%v-%v", "kube-controller-manager", g.masterName), metav1.NamespaceSystem, ports.InsecureKubeControllerManagerPort)
        if err != nil {
                return ControllerManagerMetrics{}, err
```